File: electron/patcher/thorPatcher.ts
import AdmZip from 'adm-zip';
import fs from 'fs';
import path from 'path';
import { GrfReader, GrfWriter } from './grf';

/**
 * THOR patch file structure extraction and GRF merging.
 */
export async function extractThorPatch(
    thorPath: string,
    targetDir: string,
    defaultGrfName: string
): Promise<void> {
    if (!fs.existsSync(thorPath)) {
        throw new Error(`THOR patch file not found: ${thorPath}`);
    }

    const zip = new AdmZip(thorPath);
    const entries = zip.getEntries();

    // Files destined for GRF
    const grfFiles = new Map<string, Buffer>();
    // Files destined for Disk
    const diskFiles: AdmZip.IZipEntry[] = [];

    const grfPath = path.join(targetDir, defaultGrfName);
    // Check if we should patch GRF (exists and we have files for it)
    // For now, let's always try to patch GRF if files are in 'data/' and GRF exists.
    // Otherwise fallback to disk extraction (folders).
    const useGrf = fs.existsSync(grfPath);

    for (const entry of entries) {
        if (entry.isDirectory) continue;

        const entryName = entry.entryName; // e.g. "data/xml/iteminfo.lua"

        // Check filtering logic
        // Usually only 'data/' folder content goes to GRF
        // Root files (updater.exe, dlls) go to disk
        // normalize path separators
        const normalized = entryName.replace(/\\/g, '/');

        if (useGrf && (normalized.startsWith('data/') || normalized.startsWith('data\\'))) {
            // Add to GRF batch
            // Note: entryName in GRF usually keeps 'data\...' prefix or relative?
            // Standard RO GRF contains paths like "data\texture\..."
            // So we keep the full path.
            grfFiles.set(normalized, entry.getData());
        } else {
            diskFiles.push(entry);
        }
    }

    // 1. Process Disk Files
    for (const entry of diskFiles) {
        const entryName = entry.entryName;
        // Logic for disk extraction
        const targetPath = path.join(targetDir, entryName);
        const targetDirPath = path.dirname(targetPath);

        if (!fs.existsSync(targetDirPath)) {
            fs.mkdirSync(targetDirPath, { recursive: true });
        }

        const content = entry.getData();
        fs.writeFileSync(targetPath, content);
    }

    // 2. Process GRF Files
    if (grfFiles.size > 0) {
        console.log(`Patching ${grfFiles.size} files into ${defaultGrfName}...`);

        // GRF Logic
        const reader = new GrfReader(grfPath);
        const writer = new GrfWriter();

        try {
            const header = reader.readHeader();
            const table = await reader.readFileTable(header);
            reader.close(); // Close read handle before writing

            await writer.quickMerge(grfPath, header, table, grfFiles, new Set());
        } catch (error) {
            console.error('Failed to patch GRF:', error);
            // Fallback: extract to disk if GRF fails?
            // Or throw error. Since user approved Native Patching, error is better to alert integrity issues.
            throw new Error(`Failed to patch GRF ${defaultGrfName}: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }
}

/**
 * Validate a THOR patch file
 */
export function validateThorPatch(thorPath: string): { valid: boolean; error?: string } {
    try {
        if (!fs.existsSync(thorPath)) {
            return { valid: false, error: 'File not found' };
        }

        const zip = new AdmZip(thorPath);
        const entries = zip.getEntries();

        if (entries.length === 0) {
            return { valid: false, error: 'Empty archive' };
        }

        return { valid: true };
    } catch (error) {
        return {
            valid: false,
            error: error instanceof Error ? error.message : 'Invalid archive format'
        };
    }
}

/**
 * Get list of files in a THOR patch
 */
export function getThorPatchContents(thorPath: string): string[] {
    try {
        const zip = new AdmZip(thorPath);
        return zip.getEntries()
            .filter(entry => !entry.isDirectory)
            .map(entry => entry.entryName);
    } catch {
        return [];
    }
}

File: src-tauri/src/patcher/thor_patcher.rs
use super::grf::{GrfReader, GrfWriter};
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::Path;
use zip::ZipArchive;

pub fn extract_thor_patch(
    thor_path: &str,
    target_dir: &str,
    default_grf_name: &str,
) -> Result<(), String> {
    if !Path::new(thor_path).exists() {
        return Err(format!("THOR patch file not found: {}", thor_path));
    }

    let file = fs::File::open(thor_path)
        .map_err(|e| format!("Failed to open THOR file: {}", e))?;

    let mut archive = ZipArchive::new(file)
        .map_err(|e| format!("Failed to read THOR archive: {}", e))?;

    // Separate files by destination
    let mut grf_files: HashMap<String, Vec<u8>> = HashMap::new();
    let mut disk_files: Vec<(String, Vec<u8>)> = Vec::new();

    let grf_path = Path::new(target_dir).join(default_grf_name);
    let use_grf = grf_path.exists();

    for i in 0..archive.len() {
        let mut file = archive.by_index(i)
            .map_err(|e| format!("Failed to read archive entry: {}", e))?;

        if file.is_dir() {
            continue;
        }

        let entry_name = file.name().to_string();
        let normalized = entry_name.replace('\\', "/");

        let mut data = Vec::new();
        std::io::Read::read_to_end(&mut file, &mut data)
            .map_err(|e| format!("Failed to read file data: {}", e))?;

        // Files in data/ go to GRF if it exists, otherwise to disk
        if use_grf && (normalized.starts_with("data/") || normalized.starts_with("data\\")) {
            grf_files.insert(normalized, data);
        } else {
            disk_files.push((entry_name, data));
        }
    }

    // Extract disk files
    for (entry_name, data) in disk_files {
        let target_path = Path::new(target_dir).join(&entry_name);
        
        if let Some(parent) = target_path.parent() {
            fs::create_dir_all(parent)
                .map_err(|e| format!("Failed to create directory: {}", e))?;
        }

        fs::write(&target_path, data)
            .map_err(|e| format!("Failed to write file: {}", e))?;
    }

    // Patch GRF files
    if !grf_files.is_empty() {
        println!("Patching {} files into {}...", grf_files.len(), default_grf_name);

        let reader = GrfReader::new(grf_path.to_string_lossy().to_string());
        let writer = GrfWriter::new();

        let header = reader.read_header()
            .map_err(|e| format!("Failed to read GRF header: {}", e))?;
        
        let table = reader.read_file_table(&header)
            .map_err(|e| format!("Failed to read GRF table: {}", e))?;

        writer.quick_merge(
            &grf_path.to_string_lossy(),
            &header,
            table,
            grf_files,
            HashSet::new(),
        )?;
    }

    Ok(())
}

pub fn validate_thor_patch(thor_path: &str) -> Result<bool, String> {
    if !Path::new(thor_path).exists() {
        return Err("File not found".to_string());
    }

    let file = fs::File::open(thor_path)
        .map_err(|e| format!("Failed to open file: {}", e))?;

    let archive = ZipArchive::new(file)
        .map_err(|_| "Invalid archive format".to_string())?;

    if archive.len() == 0 {
        return Err("Empty archive".to_string());
    }

    Ok(true)
}
